### PR TITLE
Signal Event Definition

### LINKED ISSUE #1161: Signals properties
Required by: https://github.com/ProcessMaker/processmaker/issues/2990

Add signal reference property to all the elements that use a signal event definition.

### PR DESCRIPTION
**Pull request by caleeli**
Resolves #1161 

This PR add the properties and definition required to configure Signal Events to the following elements:

- Start Signal Event
- Intermediate Thrown Signal Event
- Intermediate Catch Signal Event
- Boundary Catch Signal Event
- End Signal Event

Required by: https://github.com/ProcessMaker/processmaker/issues/2990

### CODE REVIEW

File: src/components/InspectorExtensionManager.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
-export default function registerInspectorExtension(node, config) {
-  const inspectorItems = getInspectorItems(node, config);
+export default function registerInspectorExtension(node, config, root = '') {
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/InspectorExtensionManager.js:1]: Yes, it do not require a default value. Default value removed.

File: src/components/InspectorExtensionManager.js [hunk 2]
```diff
@@ -27,11 +28,21 @@ function addInspectorItem(inspectorItems, config) {
...
+    node = node.inspectorConfig[0];
+    root = root.split('.');
+    while (node && root.length) {
+      node = node.items.find(node => node.config && node.config.name === root[0]);
+      root.shift();
+    }
+    return node ? node.items : null;
+  } else {
+    return node.inspectorConfig[0].items[0].items;
+  }
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/InspectorExtensionManager.js]: Refactored

File: src/components/inspectors/InspectorPanel.vue [hunk 3]
```diff
@@ -177,7 +177,7 @@ export default {
       return definition.targetRef.$type === 'bpmn:CallActivity';
     },
     customInspectorHandler(value) {
-      return this.nodeRegistry[this.highlightedNode.type].inspectorHandler(value, this.highlightedNode, this.setNodeProp, this.moddle);
+      return this.nodeRegistry[this.highlightedNode.type].inspectorHandler(value, this.highlightedNode, this.setNodeProp, this.moddle, this.$parent.definitions);
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/inspectors/InspectorPanel.vue:180]: Added property `definitions` to InspectorPanel

File: src/components/inspectors/advancedAccordionConfig.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ export default {
     initiallyOpen: false,
     label: 'Advanced',
     icon: 'cogs',
-    name: 'inspector-accordion',
+    name: 'advanced-accordion',
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/inspectors/advancedAccordionConfig.js]: Reverted. Any way I think the `name` should be unique, to follow the same behavior of the items for which the `name` is used as the component `id` (rendered by ScreenRenderer)

File: src/components/nodes/signalStartEvent/signalStartEvent.vue [hunk 5]
```diff
@@ -4,6 +4,7 @@ import signalStartEventIcon from '@/assets/signal-start-event.svg';
 
 export default {
   extends: startEvent,
+  props: ['moddle', 'rootElements', 'id'],
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/nodes/signalStartEvent/signalStartEvent.vue:8]: Removed

File: src/components/nodes/signalEventDefinition/index.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    let signal = definitions.rootElements.find(element => element.id === value.signalRef);
+    if (!signal) {
+      signal = moddle.create('bpmn:Signal', {
+        id: value.signalRef,
+        name: value.signalRef,
+      });
+      definitions.rootElements.push(signal);
+    }
+    node.definition.get('eventDefinitions')[0].signalRef = signal;
+    this.cleanSignalDefinitions(definitions);
```
[the remark this answers is not in the document]
Comment (caleeli) [reply src/components/nodes/signalEventDefinition/index.js:33]: No, the user could use the same signal for multiple elements in the same diagram, then if the user changes the signal for one, the old signal should not be removed if it is referenced in other elements.

**Comment by chipit24** [comment]
@caleeli I went through the functionality and have a couple questions:

1. The Signal End Event is also a throwing event, but I don't see the Signal Ref property for it. Should it also have it?
   Screen Shot 2020-04-02 at 11 47 30 AM

2. The Signal Start Event and Intermediate Signal Catch Event are both catching event, but I don't see a way to select which signals they should listen for.
   Screen Shot 2020-04-02 at 11 48 32 AM
The catching message events have a select input for this, should the same be added for signals?
   Screen Shot 2020-04-02 at 11 48 45 AM

**Comment by caleeli** [comment]
I completed by adding the missing signal references for those elements.

Similar to the **Screen Selection Control** in Tasks, I implemented the **Signal Selection Control** in the ProcessMaker project, that control uses an API to get a list of all the signals defined in other processes. Since this is an extension to the standard (to use signals defined in others process definitions), I preferred to keep the signal reference in the Modeler project as an input text to configure the signal ID to catch or launch.

**Pull request merged**